### ISSUE TITLE
Get Coordinates of the target Player.

### ISSUE THREAD

**Issue by Azatik** [SpongePowered/SpongeAPI#365]
Get Coordinates of the target Player. There are similar in Bukkit:
Block block = player.getTargetBlock(null, 100);

**Comment by gabizou** [comment]
Can you be a little more descriptive here?

**Comment by Azatik** [comment]
It is very difficult. Badly I know English.

**Comment by Sannholm** [comment]
I'm guessing he wants to have access to block/entity raytracing methods

**Comment by Azatik** [comment]
Screenshot:
![2015-01-04_19 30 07](https://cloud.githubusercontent.com/assets/10154956/5611286/841d8022-94e6-11e4-8346-e9adc138a7c7.png)

**Comment by ryantheleach** [comment]
Do you want it to work with blocks at any distance, or just blocks within the default minecraft reach?

**Comment by Azatik** [comment]
Just get the data on this block.

**Comment by ryantheleach** [comment]
I don't see the advantage of using an iterator for finding when the LOS should stop, Can't you do anything you would with an interator with a predicate and more?

The only issue I see is if people forget to count the number of uses inside the predicate they could search forever, and I'm not 100% sure what would happen if the predicate had state that was changing in parallel.

**Comment by nightpool** [comment]
@ryantheleach I agree, but I think the argument for the predicate is that it can more elegantly accomplish all of the things that the 4 arguments do, with more flexibility, and that any downsides to the API (people forgetting to count things, for example) can overcome by providing a few pre-built default helpers (static BlockPredicate.distance(Location start, int radius); for example)

**Comment by nightpool** [comment]
Some helpers that occur off the top of my head:

````java
distance(Location start, int radius); // stop after a certain distance
blockType(BlockType type); // stop when you hit a certain type of block
whenSolid(Predicate<BlockType> transparent); // stop when you hit a solid block
```

**Comment by Sannholm** [comment]
How about proper entity raytracing? It is one of the things I've really been missing in Bukkit.

**Comment by Deamon5550** [comment]
The thing with these types of utilities is that there is not real advantage to building them into the API over writing them yourself. I would argue that for the sake of not bloating the API with functionality that can just as easily be created by a plugin, with no performance or memory costs over the API implementation, that it should be left to plugins to implement.

**Comment by Axiometry** [comment]
@Deamon5550 Raytracing implies knowledge of bounding boxes and solidity, among other block properties and game logic. Recreating *true* raytracing in a plugin would be quite an exercise in reinventing the wheel, whereas integrating it into the API allows it to be implementation-specific. How would mod blocks be handleable otherwise?

**Comment by Azatik** [comment]
I consider it useful. Especially, if events are occupied: RightClick and LeftClick.
Application: working with blocks and entities in the distance.

**Comment by boformer** [comment]
I agree with @Azatik.

**Comment by Azatik** [comment]
I think the addition of two methods in the interface player.
getTargetLocation: get the Coordinates of where the player is looking.
getTargetEntity: get the Entity of which looks player.

**Comment by simon816** [comment]
I can see `getTargetLocation` being useful. `getTargetEntity` is too specific, perhaps iterate over entities within certain bounds (something like targetLoc + 1, targetLoc - 1)
If a player is looking at a block, it would also be useful to get the face of the block they're looking at.
i.e. a tuple of (Location, Direction)

**Comment by ST-DDT** [comment]
Maybe we should move the second method (`getTargetEntity`) to a util class. With some customization options.

**Comment by ST-DDT** [comment]
Here is the Gist providing the abstract functionality for target block / line of sight functionality.

https://gist.github.com/ST-DDT/e49fc174f19a65f26b1b

Shall i create a PR for that?

**Comment by ST-DDT** [comment]
I created a PR to fix this issue:
https://github.com/SpongePowered/SpongeAPI/pull/683

**Comment by ryantheleach** [comment]
Should this PR include entity tracing?

**Comment by gratimax** [comment]
Fixed in #687

**Comment by DDoS** [comment]
Just a note: until the EyeLocationData is implemented, the block tracer will use the player's location (feet). This means that the result is a bit lower than expected.
